### ISSUE TITLE
Issue with gradients wrt optimality fn parameters through root finding vjp

### ISSUE THREAD

**Issue by EiffL** [google/jaxopt#141]
First of all, thanks a lot for this library! Really useful tools! 
I'm interested in getting at least 2nd order gradients through root finding, and I'm finding an odd behavior that I wanted to report. 

Maybe I'm doing something wrong, but in the following schematic case I silently get the wrong gradients:
```python
def inv_f(x, aux):
  bisec = Bisection(optimality_fun=F, lower=0.0, upper=1., 
                    check_bracket=False, unroll=True)
  return bisec.run(aux=aux).params

# Here I extract the value part of the vjp, but the grad part also gives wrong results
test_fn = lambda aux: jax.value_and_grad(inv_f)(0.5, aux)[0] 

jax.grad(test_fn)(1.) # Returns 0 instead of the expected gradients
```
Here I'm only trying to get gradients of the value returned by `jax.value_and_grad`, but the gradients of the gradients returned by `jax.value_and_grad` are also wrong (but not as obvious). 

I made a small **demo notebook that reproduces this issue [here](https://colab.research.google.com/drive/1D5FHJJFyXmFuCS45jUjHJvxIQUhBemYy?usp=sharing)**. 

As a reference I've also implemented my own implicit gradients, bypassing the jaxopt ones, and they seem to give me the correct answer.

Reading the source code of jaxopt, it is not immediatly obvious to me why this doesn't work... Sorry I couldn't directly suggest a PR, but I hope this report is still useful (and that I'm not just using jaxopt wrong).

**Comment by mblondel** [comment]
Hello @EiffL! 

I haven't checked the notebook but in the code snippet above there are two issues: `x` is not used and if you want to differentiate with respect to `aux` (the second argument) you need to specify `argnums=1` in `value_and_grad`. Also I would not set set `unroll=True`.

There is a code [example](https://jaxopt.github.io/stable/root_finding.html) in the documentation on how to differentiate bisection. Here's how to add second-order derivatives:
```python
import jax
from jaxopt import Bisection

def F(x, factor):
  return factor * x ** 3 - x - 2

def root(factor):
  bisec = Bisection(optimality_fun=F, lower=1, upper=2)
  return bisec.run(factor=factor).params

first_deriv = jax.grad(root)
second_deriv = jax.grad(first_deriv)

def fd(factor, eps=1e-4):
  """Compute first-order derivative by finite difference"""
  return (root(factor + eps) - root(factor - eps)) / (2 * eps)

def fd2(factor, eps=1e-4):
  """Compute second-order derivative by finite difference"""
  return (first_deriv(factor + eps) - first_deriv(factor - eps)) / (2 * eps)

factor = 2.0

print("First-order derivative")
print(first_deriv(factor))
print(fd(factor))
print()

print("Second-order derivative")
print(second_deriv(factor))
print(fd2(factor))
print()
```
It's actually quite cool that this works because I don't think we have any tests for second-order derivatives yet!

**Comment by EiffL** [comment]
Thanks so much for the quik feedback! Ah, my attempt at simplifiction actually ended up overly-simplified/wrong >.< 

Yes the example from the documentation works nicely, and even to higher order :-) 

But it's a slightly different use case that doesn't work for me, I first take the gradient with respect to the root, and then I want to take the gradient of that with respect to the function parameters. 

Here is the actual code from the notebook:
```python
# Some function in [0,1] we want to inverse
f = lambda x,a,b: (a*x**2 + b*x) /(a + b )

# Optimality function
def F(x, aux, y):
  a,b = aux
  return f(x,a,b) - y

# Inverse function using jaxopt directly
def inv_f(y, params):
  a,b =params
  bisec = Bisection(optimality_fun=F, lower=0.0, upper=1., 
                    check_bracket=False, unroll=True)
  return bisec.run(aux=(a,b),y=y).params

# Now I define a function of the paramaters
def myfunc(a):
  params = [jnp.squeeze(a), 0.5]
  return jax.value_and_grad(inv_f)(0.5, params)[0]

jax.grad(myfunc)(2.) # Gives wrong answer
```

**Comment by mblondel** [comment]
Why do you use `jax.value_and_grad(inv_f)(0.5, params)[0]` instead of `inv_f(0.5, params)`? If you return `[0]` only, this doesn't need the gradient. Strangely, I get the correct result when using the latter but not the former:
```python
def myfunc(a):
  params = [jnp.squeeze(a), 0.5]
  return jax.value_and_grad(inv_f)(0.5, params)[0]

def myfunc2(a):
  params = [jnp.squeeze(a), 0.5]
  return inv_f(0.5, params)

def fd(a, eps=1e-2):
  return (myfunc(a + eps) - myfunc(a - eps)) / (2 * eps)

def fd2(a, eps=1e-2):
  return (myfunc2(a + eps) - myfunc2(a - eps)) / (2 * eps)

print(jax.grad(myfunc)(2.)) # Gives wrong answer
print(fd(2.0))
print()

print(jax.grad(myfunc2)(2.))
print(fd2(2.0))
```
This prints
```
0.0
0.01373291

0.0136953145
0.01373291
```

By the way, finite differences can be very sensitive to the choice of `eps` in float32 precision. I usually try 1e-2, 1e-3, 1e-4.

@froystig Any idea why `jax.grad` gives the wrong derivative when using `jax.value_and_grad(inv_f)(...)[0]` instead of `inv_f(...)`?

**Comment by EiffL** [comment]
yep, that's the issue. The difference between taking the gradients of `jax.value_and_grad(inv_f)(...)[0]` instead of `inv_f(...)` is that internally jax will execute the forward function through a vjp:
https://github.com/google/jax/blob/113cd9b939afb7516af0bdf34b1499b271201111/jax/_src/api.py#L1063
So, instead of executing the root finding function directly, it will use the forward_fn of the custom vjp, and I guess something must be blocking the gradients in there.

I'm guessing something weird must be happening in the tracking of additional function arguments in the custom vjp but couldn't track down where the problem happens.

I don't actually care that much about the gradients through `jax.value_and_grad(inv_f)(0.5, params)[0]` and I need the gradients through `jax.value_and_grad(inv_f)(0.5, params)[1]`, but both are actually wrong,  it's just more evident with the grad of the function.

If I wrap the bisection call in my own implicit grad, it works as expected.
```python
import jax
import jax.numpy as jnp
from functools import partial
from jaxopt import Bisection
from jaxopt.linear_solve import solve_normal_cg

# Some function in [0,1] we want to inverse
f = lambda x,a,b: (a*x**2 + b*x) /(a + b )

# Optimality function
def F(x, aux, y):
  a,b = aux
  return f(x,a,b) - y

# Inverse function using jaxopt directly
def inv_f(y, params):
  a,b =params
  bisec = Bisection(optimality_fun=F, lower=0.0, upper=1.,
                    check_bracket=False, unroll=True)
  return bisec.run(aux=(a,b),y=y).params

@partial(jax.custom_vjp, nondiff_argnums=(0,))
def fixed_point_layer(f, params):
  a,b, y =params
  bisec = Bisection(optimality_fun=f, lower=0.0, upper=1., 
                    check_bracket=False, unroll=True)
  return bisec.run(aux=(a,b),y=y).params

def fixed_point_layer_fwd(f, params):
  z_star = fixed_point_layer(f, params)
  return z_star, (params, z_star)

def fixed_point_layer_bwd(f, res, z_star_bar):
  x, z_star = res
  def internal_f(params, z):
    a,b,y = params
    return f(z, (a,b),  y)
  _, vjp_a = jax.vjp(lambda x: internal_f(x, z_star), x)
  _, vjp_z = jax.vjp(lambda z: internal_f(x, z), z_star)
  return vjp_a(solve_normal_cg(lambda u: vjp_z(u)[0], - z_star_bar))
  
fixed_point_layer.defvjp(fixed_point_layer_fwd, fixed_point_layer_bwd)

def inv_f2(x, params):
  return fixed_point_layer(F, params+[x])

def myfunc2(a):
  params = [jnp.squeeze(a), 0.5]
  return jax.value_and_grad(inv_f2)(0.5, params)[0]

jax.grad(myfunc2)(2.)  # Returns 0.01369531
```

**Comment by mblondel** [comment]
It seems like it's indeed a bug in our implicit diff machinery then. Thanks a lot!

**Comment by froystig** [comment]
This was a helpful discovery, since the issue here was actually affecting all solvers, not only bisection. I commented in #143 about why it was more apparent when using bisection in particular. Thank you for reporting this!

**Comment by EiffL** [comment]
Many thanks for the quick fix, and happy it turned out to be a useful find. And thanks again for your work on this package!
